### ISSUE TITLE
Piecewise regression with multidimensional arrays with extra 'otherwise' condition

### ISSUE THREAD

**Issue by bastula** [numpy/numpy#5737]
Comments from (#4792):

The changes applied from this pull request don't seem to work well with multidimensional arrays that have an extra "otherwise" condition:

```python
>>> x = array([[-2.5, -1.5, -0.5],
               [ 0.5,  1.5,  2.5]])
>>> np.piecewise(x, [x < 0, x >= 2], [-1, 1, 3])
ValueError: all the input arrays must have same number of dimensions
```
The same code produces the following result in NumPy 1.8.2:

```python
>>> array([[-1., -1., -1.],
           [ 3.,  3.,  1.]])
```
If you reshape the array to a single dimension, it works as intended.

Also, I noticed there aren't any multidimensional test cases, so maybe this can be one of them?

These conditions were tested with Python 2.7 / NumPy 1.8.2 and Python 2.7.9 / NumPy 1.9.2.

**Comment by bastula** [comment]
Any thoughts on this? I had to revert the code in ```numpy/lib/function_base.py``` from #4792 and it worked correctly.

**Comment by njsmith** [comment]
It does sound like a bug. If you want to have a go at fixing it then pull
requests requests are gratefully accepted :-)

**Comment by bastula** [comment]
The easiest way would be to revert the #4792 as not only does this issue arise, but #5729 was caused as a result.

I would need to spend more time to see if we can resolve all 3 issues in one fix.

**Comment by charris** [comment]
Would be good to get this fixed before 1.10 branch.

**Comment by badray** [comment]
+1
Had the same issue with similar code. Reverting to 1.8.0 casues my code to work as expected.

**Comment by charris** [comment]
I'm pushing this off to 1.11. If anyone wants to submit a fix pronto it will be looked at.

**Comment by bastula** [comment]
What's the ETA on the 1.10 release?

**Comment by bastula** [comment]
Closed by PR #7145.
